packages/mql-typescript/src/driverSchema/docsCrawler.ts

```diff
@@ -0,0 +1,387 @@
+import JSON5 from 'json5';
+import { removeNewlines, removeTrailingComments } from '../utils';
+import { getSimplifiedSchema } from 'mongodb-schema';
+import type { SimplifiedSchema } from 'mongodb-schema';
+import { JSDOM, VirtualConsole } from 'jsdom';
+import { createHash } from 'crypto';
+import { BSON } from 'bson';
+
+abstract class CustomTypeProcessor {
+  public abstract process(json: string): string;
+
+  public abstract canRevive(value: any): boolean;
+
+  public abstract revive(value: any): any;
+}
+
+abstract class RegexCustomTypeProcessor extends CustomTypeProcessor {
+  private regex: RegExp;
+  private replacementPrefix: string;
+  protected abstract reviveCore(value: string): any;
+
+  protected constructor(regex: string) {
+    super();
+    this.regex = new RegExp(regex, 'g');
+    this.replacementPrefix =
+      '!rctp-' + createHash('sha256').update(regex).digest('base64');
+  }
+
+  public process(json: string): string {
+    return json.replace(this.regex, `"${this.replacementPrefix}$1"`);
+  }
+
+  public canRevive(value: any): boolean {
+    return (
+      typeof value === 'string' && value.startsWith(this.replacementPrefix)
+    );
+  }
+
+  public revive(value: string): any {
+    return this.reviveCore(value.replace(this.replacementPrefix, ''));
+  }
+}
+
+class IsoDateProcessor extends RegexCustomTypeProcessor {
+  constructor() {
+    super('\\bISODate\\(\\s*"([^"]*)"\\s*\\)');
+  }
+
+  public reviveCore(value: string): Date | undefined {
+    const isoDateRegex =
+      /^(?<Y>\d{4})-?(?<M>\d{2})-?(?<D>\d{2})([T ](?<h>\d{2})(:?(?<m>\d{2})(:?((?<s>\d{2})(\.(?<ms>\d+))?))?)?(?<tz>Z|([+-])(\d{2}):?(\d{2})?)?)?$/;
+    const match = isoDateRegex.exec(value);
+    if (match !== null && match.groups !== undefined) {
+      // Normalize the representation because ISO-8601 accepts e.g.
+      // '20201002T102950Z' without : and -, but `new Date()` does not.
+      const { Y, M, D, h, m, s, ms, tz } = match.groups;
+      const normalized = `${Y}-${M}-${D}T${h || '00'}:${m || '00'}:${
+        s || '00'
+      }.${ms || '000'}${tz || 'Z'}`;
+      const date = new Date(normalized);
+      // Make sure we're in the range 0000-01-01T00:00:00.000Z - 9999-12-31T23:59:59.999Z
+      if (
+        date.getTime() >= -62167219200000 &&
+        date.getTime() <= 253402300799999
+      ) {
+        return date;
+      }
+    }
+
+    return new Date(value);
+  }
+}
```

Comment:
Yeah, this looks a lot like the type of logic we have in https://github.com/mongodb-js/devtools-shared/tree/main/packages/ejson-shell-parser 🙂